applications/TopologyOptimizationApplication/custom_utilities/structure_response_function_utilities.h

```diff
@@ -0,0 +1,297 @@
+{
+public:
+
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of StructureResponseFunctionUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(StructureResponseFunctionUtilities);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    StructureResponseFunctionUtilities( ModelPart& model_part )
+    : mr_structure_model_part(model_part)
+    {
+    }
+
+    /// Destructor.
+    virtual ~StructureResponseFunctionUtilities()
+    {
+    }
+
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    // ---------------------------------------------------------------------------------------------------------------------------------------------
+    // --------------------------------- COMPUTE STRAIN ENERGY -------------------------------------------------------------------------------------
+    // ---------------------------------------------------------------------------------------------------------------------------------------------
+
+    /// Computes the strain energy as the objective function of the optimization problem.
+    double ComputeStrainEnergy()
+    {
+        KRATOS_TRY;
+
+        clock_t begin = clock();
+        std::cout<<"  Start calculating strain energy."<<std::endl;
+    /* 	std::cout<< "respones function bevor Schleife:" << std::endl;  */
+
+        double Out = 0.0;
+        double Global_Strain_Energy = 0.0;
+
+        // Loop over all elements to calculate their local objective function and sum it into the global objective function (Global Strain Energy)
+        for( ModelPart::ElementIterator element_i = mr_structure_model_part.ElementsBegin(); element_i!= mr_structure_model_part.ElementsEnd();
+                element_i++ )
+        {
+
+            element_i->Calculate(LOCAL_STRAIN_ENERGY, Out, mr_structure_model_part.GetProcessInfo());
+
+            Global_Strain_Energy += element_i->GetValue(LOCAL_STRAIN_ENERGY);
+            
+        }
+
+        clock_t end = clock();
+        std::cout << "  Strain energy calculated                  [ spent time =  " << double(end - begin) / CLOCKS_PER_SEC << " ] " << std::endl;
+
+        // Return this obtained Global Strain Energy value as the objective function of the complete system
+        return Global_Strain_Energy;
+
+        KRATOS_CATCH("");
+    }
+
+    double ComputeVolumeFraction()
+    {
+        KRATOS_TRY;
+
+        clock_t begin = clock();
+        std::cout<<"  Start calculating volume fraction."<<std::endl;
+
+        int number_elements = 0;
+        double Global_Volume_Fraction = 0.0;
+
+
+        // Loop over all elements to obtain their X_PHYS and know how many elements the model has
+        for( ModelPart::ElementIterator element_i = mr_structure_model_part.ElementsBegin(); element_i!= mr_structure_model_part.ElementsEnd();
+                element_i++ )
+        {
+            Global_Volume_Fraction += element_i->GetValue(X_PHYS);
```

Comment:
Yes exactly. This is the same for the constraint sensitivity, the way it is implemented at teh moment it is only valid when the elements all have the same size! How can I implement the volume of the element?

Reply:
Each element geometry has a function "element.GetGeometry().DomainSize()" that gives you the Volume/Area/length of the element depending on the dimension. 

You can simply multiply your volume_fraction value with the element volume to get something that takes the element size into account.

Reply:
I implemented within the loop, where the optimization runs trough the elements with 
`elemental_volume = element_i->GetGeometry().DomainSize();`
I called the element volume. But something goes wrong here, as I get results from `-8.46499e+09` up to `1.35694e+09` for the element size. Did I implement it wrong or is this the wrong function for what we intend to do?

Reply:
As the element size is so high, the resulting volume fraction is `-5.24138e+07`, whereas it should be `0.5`

Reply:
This is very strange, no idea why the values you get are so large in magnitude and some even negative, that seems to be wrong indeed. 

In any case you would not get 0.5 as response value but 0.5*total volume. 

Anyways, i think this can be done in a second step as well. Since what you have implemented is correct, its just not what i would expect by the name of the function.

Reply:
Could it be that the way I called the function is wrong or gives a different value as expected?
Yes you are right. This only works as long as the elements have the same size.

Reply:
This is how it is used in ShapeOpt to compute the volume, there is nothing special in how it is called. 

In the small cantilever example you should get element volumes in the range of 1e-7 to 1e-6.  Not sure why the values are so strange in your case. Ev. they are computed in the deformed configuration (displacement of the struct. analysis) - But unless you have very unphysically large displacements (e.g. due to unit error in the material/loads etc) this should still give reasonable values.

Reply:
This could be, that this computed in the deformed configuration. Due to the fact, that the youngs modulus is set to 1, the deformation is quite large! Can I call the undeformed element size?

Reply:
That might explain it. I am not aware of a method to compute the volume of the undeformed geometry. I see the following options:
- temporarily set the coordinates of the geometry to the initial state before computing the volume and reset afterwards. 
- implement your own volume function (not nice if you have different element types)

Alternatively you could do this in the meantime (but its just to check if its because of displacements): Use a small load that results in a reasonable displacement (< length/10) or increase the Young's modulus. (for the optimization problem it should not change anything since its linear static)

Reply:
You were right! When increasing the youngs modulus to 10000000, the element size is as you expected: around 1e-6. I will try to implement your suggestion and set the coordinates to the initial state. 

Can I do that with using: GetInitialPosition()? 

I would do that locally though, I would leave this as it is now, if that is not a problem?!

Reply:
Yeah, this is perfectly suited for a follow up PR.

Reply:
I was able to implement this, so that the element volume is considered!